ydb/core/tx/schemeshard/schemeshard_info_types.cpp

```diff
@@ -46,6 +80,111 @@ void TSubDomainInfo::ApplyAuditSettings(const TSubDomainInfo::TMaybeAuditSetting
     }
 }
 
+TDiskSpaceQuotas TSubDomainInfo::GetDiskSpaceQuotas() const {
+    if (!DatabaseQuotas) {
+        return {};
+    }
+
+    ui64 hardQuota = DatabaseQuotas->data_size_hard_quota();
+    ui64 softQuota = DatabaseQuotas->data_size_soft_quota();
+
+    if (!softQuota) {
+        softQuota = hardQuota;
+    } else if (!hardQuota) {
+        hardQuota = softQuota;
+    }
+
+    THashMap<TString, TQuotasPair> storagePoolsQuotas;
+    for (const auto& storagePoolQuota : DatabaseQuotas->storage_pools_quotas()) {
+        ui64 storagePoolHardQuota = storagePoolQuota.data_size_hard_quota();
+        ui64 storagePoolSoftQuota = storagePoolQuota.data_size_soft_quota();
+
+        if (!storagePoolSoftQuota) {
+            storagePoolSoftQuota = storagePoolHardQuota;
+        } else if (!storagePoolHardQuota) {
+            storagePoolHardQuota = storagePoolSoftQuota;
+        }
+
+        storagePoolsQuotas.emplace(storagePoolQuota.storage_pool_kind(), TQuotasPair{
+                .HardQuota = storagePoolHardQuota,
+                .SoftQuota = storagePoolSoftQuota
+            }
+        );
+    }
+
+    return TDiskSpaceQuotas{hardQuota, softQuota, std::move(storagePoolsQuotas)};
+}
+
+bool TSubDomainInfo::CheckDiskSpaceQuotas(IQuotaCounters* counters) {
+    const auto changeSubdomainState = [&](EDiskUsageStatus diskUsage) {
+        if (diskUsage == EDiskUsageStatus::AboveHardQuota && !DiskQuotaExceeded) {
+            counters->ChangeDiskSpaceQuotaExceeded(+1);
+            DiskQuotaExceeded = true;
+            ++DomainStateVersion;
+            return true;
+        }
+        if (diskUsage == EDiskUsageStatus::BelowSoftQuota && DiskQuotaExceeded) {
+            counters->ChangeDiskSpaceQuotaExceeded(-1);
+            DiskQuotaExceeded = false;
+            ++DomainStateVersion;
+            return true;
+        }
+        return false;
+    };
+
+    auto quotas = GetDiskSpaceQuotas();
+    if (!quotas) {
+        return changeSubdomainState(EDiskUsageStatus::BelowSoftQuota);
+    }
+
+    ui64 totalUsage = TotalDiskSpaceUsage();
+    const auto storagePoolsUsageStatus = CheckStoragePoolsQuotas(DiskSpaceUsage.Tables.StoragePoolsUsage, quotas.StoragePoolsQuotas);
+    if (quotas.HardQuota && totalUsage > quotas.HardQuota
```

Comment:
I hope this condition is easy to understand. In other words, we want to account for both:
1) quotas.HardQuota
and
2) storage pools quotas.
If any one of them is present, we check, if it was exceeded.

Reply:
Changed the code to be more explicit and added a comment.